### PR TITLE
Birectional communication of libdnf5 actions plugin with running processes - "json" mode

### PR DESCRIPTION
**Pull request by jrohel**
Closes: https://github.com/rpm-software-management/dnf5/issues/1538

Introduces a "mode" option in the hook handler definition.
Option "mode" defines communication model between actions plugin and process.
Introduces 2 modes of communication: "plain" and "json".

### "mode=plain" - Original and default communication model.

Parameters are passed to the process only by arguments when it starts. Arguments may contain variables that are substituted before the process is started. A process can modify libdnf5 settings (main configuration, repository configuration, variables) and plugin actions variables by writing to standard
output.

**Output line format:**
- `tmp.<actions_plugin_variable_name>=<value>` - sets the value of action plugins variable `<actions_plugin_variable_name>`
-` tmp.<actions_plugin_variable_name>` - removes the action plugins variable if it exists
- `conf.<option_name>=<value>` - sets the value of option <option_name> in the base configuration
- `conf.<repoid_pattern>.<option_name>=<value>` - sets the value of option <option_name> in the matching repositories
- `var.<variable_name>=<value>` - sets value of the variable <variable_name>

### "mode=json" - Newly introduced communication model.

Parameters can be passed to the process before it starts using arguments just like in "plain" mode. This mode adds bidirectional communication between the actions plugin and the process. This is request/response communication.
The process sends requests and the actions plugin sends responses. The process writes requests to the standard output and reads responses from the standard input.

**Processes send requests and the actions plugin responds:**

- Request format:
    `{"op":"<operation>", "domain":"<domain>", "args":{<op_domain_spec_args>}}`

"domain" is ignored in the case of the "log" operation.

- Response format:
    `{"op": "reply", "requested_op":"<op_from_request>", "domain": "<domain_from_request>", "status": "OK", "return":{<data>}}`

    `{"op": "reply", "requested_op":"<op_from_request>", "domain": "<domain_from_request>", "status": "ERROR", "message":"<error message>"`}

"return" is not included in the response to a request for a "log"
operation.

If the JSON request is unparsable - a syntax error or incomplete message, or the response cannot be written, the action plugin closes both communication pipes. The process must terminate if the communication pipes are closed.

### CODE REVIEW

File: libdnf5-plugins/actions/actions.cpp [hunk 1]
```diff
@@ -844,7 +854,777 @@ void Actions::process_command_output_line(std::string_view line) {
...
+
+                libdnf5::sack::ExcludeFlags query_flags{0};
+                auto * jparams = get_any_object_or_null(jargs, "params");
+                if (jparams) {
+                    get_array(jargs, "params");  // check, must be array
+                    for (size_t idx = 0; idx < json_object_array_length(jparams); ++idx) {
+                        auto * jparam = json_object_array_get_idx(jparams, idx);
+                        const auto param_key = get_string_view(jparam, "key");
+                        //auto * jvalue = get_any_object_or_null(jparam, "value");
+                        //const auto value = jvalue ? std::string(get_string_view(jvalue)) : "";
```
Comment (kontura) [inline libdnf5-plugins/actions/actions.cpp:1319]: These can be removed right?

File: libdnf5-plugins/actions/actions.cpp [hunk 2]
```diff
@@ -844,7 +854,777 @@ void Actions::process_command_output_line(std::string_view line) {
...
+                        }
+                    }
+                }
+
+                auto * jfilters = get_any_object_or_null(jargs, "filters");
+                if (jfilters) {
+                    get_array(jargs, "filters");  // check, must be array
+                }
+
+                enum class Direction { IN, OUT, ALL } edirection = Direction::ALL;
```
Comment (kontura) [inline libdnf5-plugins/actions/actions.cpp:1341]: Why isn't the `Action::Direction` used here?

File: libdnf5-plugins/actions/actions.cpp [hunk 3]
```diff
@@ -844,7 +854,777 @@ void Actions::process_command_output_line(std::string_view line) {
...
+                    for (size_t idx = 0; idx < json_object_array_length(jfilters); ++idx) {
+                        auto * jfilter = json_object_array_get_idx(jfilters, idx);
+                        const auto filter_key = get_string_view(jfilter, "key");
+                        auto * jvalue = get_any_object_or_null(jfilter, "value");
+                        const auto value = jvalue ? std::string(get_string_view(jvalue)) : "";
+                        auto * joperator = get_any_object_or_null(jfilter, "operator");
+                        const auto oper = joperator ? cmp_operator_from_string(get_string_view(joperator))
+                                                    : libdnf5::sack::QueryCmp::EQ;
+                        if (filter_key == "direction") {
+                        } else if (filter_key == "name") {
```
Comment (kontura) [inline libdnf5-plugins/actions/actions.cpp:1381]: Minor suggestion but perhaps we could add a comment here saying that the `direction` filtering was already done above.

**Comment by kontura** [comment]
I believe a docs update is needed and I assume it will be done in a follow up PR.

File: libdnf5-plugins/actions/actions.cpp (hunk 1 above)
Comment (jrohel) [reply libdnf5-plugins/actions/actions.cpp]: Yes. I removed it.

File: libdnf5-plugins/actions/actions.cpp (hunk 2 above)
Comment (jrohel) [reply libdnf5-plugins/actions/actions.cpp]: Thank you. Good point. This is a good candidate for using the `Action::Direction` type. I somehow missed it when I reorganized the code.

File: libdnf5-plugins/actions/actions.cpp (hunk 3 above)
Comment (jrohel) [reply libdnf5-plugins/actions/actions.cpp]: I added comments.

**Pull request merged**